accord-core/src/main/java/accord/primitives/Deps.java

```diff
@@ -0,0 +1,978 @@
+        while (index < end)
+            forEach.accept(txnIds[keyToTxnId[index++]]);
+    }
+
+    public Keys keys()
+    {
+        return keys;
+    }
+
+    public int txnIdCount()
+    {
+        return txnIds.length;
+    }
+
+    public TxnId txnId(int i)
+    {
+        return txnIds[i];
+    }
+
+    @Override
+    public Iterator<Map.Entry<Key, TxnId>> iterator()
+    {
+        return new Iterator<>()
+        {
+            int i = keys.size(), k = 0;
+
+            @Override
+            public boolean hasNext()
+            {
+                return i < keyToTxnId.length;
+            }
+
+            @Override
+            public Map.Entry<Key, TxnId> next()
+            {
+                Entry result = new Entry(keys.get(k), txnIds[keyToTxnId[i++]]);
+                if (i == keyToTxnId[k])
+                    ++k;
+                return result;
+            }
+        };
+    }
+
+    @Override
+    public String toString()
+    {
+        if (keys.isEmpty())
+            return "{}";
+
+        StringBuilder builder = new StringBuilder("{");
+        for (int k = 0, t = keys.size(); k < keys.size() ; ++k)
+        {
+            if (builder.length() > 1)
+                builder.append(", ");
+
+            builder.append(keys.get(k));
+            builder.append(":[");
+            boolean first = true;
+            while (t < keyToTxnId[k])
+            {
+                if (first) first = false;
+                else builder.append(", ");
+                builder.append(txnIds[keyToTxnId[t++]]);
+            }
+            builder.append("]");
+        }
+        builder.append("}");
+        return builder.toString();
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        return equals((Deps) o);
+    }
+
+    public boolean equals(Deps that)
+    {
+        return this.txnIds.length == that.txnIds.length
+               && this.keys.size() == that.keys.size()
+               && Arrays.equals(this.keyToTxnId, that.keyToTxnId)
+               && Arrays.equals(this.txnIds, that.txnIds)
+               && this.keys.equals(that.keys);
+    }
+
+    public static class Entry implements Map.Entry<Key, TxnId>
+    {
+        final Key key;
+        final TxnId txnId;
+
+        public Entry(Key key, TxnId txnId)
+        {
+            this.key = key;
+            this.txnId = txnId;
+        }
+
+        @Override
+        public Key getKey()
+        {
+            return key;
+        }
+
+        @Override
+        public TxnId getValue()
+        {
+            return txnId;
+        }
+
+        @Override
+        public TxnId setValue(TxnId value)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public String toString()
+        {
+            return key + "->" + txnId;
+        }
+    }
+
+    public boolean checkValid()
```

Comment:
Pushed a commit https://github.com/dcapwell/cassandra-accord/commit/3d17394952e2d498b190ab544c997a7fe9de3fbe.

I feel that this method should be what throws as it could produce a much more meaningful error than the caller can.